Remove legacy qasm2 parser

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit removes the legacy qasm2 parser. This was the original qasm parser in qiskit dating back to the very beginning of the project, but it has been superseded in recent releases by the faster more correct/strict rust parser. We no longer need to keep around two parsers for qasm2. This commit removes the legacy one and it's associated functions (like the ast converter).

### Details and comments

mtreinish committed:
Remove legacy qasm2 parser

This commit removes the legacy qasm2 parser. This was the original
qasm parser in qiskit dating back to the very beginning of the project,
but it has been superseded in recent releases by the faster more
correct/strict rust parser. We no longer need to keep around two parsers
for qasm2. This commit removes the legacy one and it's associated
functions (like the ast converter).

mtreinish committed:
Remove api docs too

mtreinish committed:
Remove ast converter benchmark

mtreinish commented:
The open question for me here is what we want to do about the pygments formatter? (Which is the source of the lint failures). Do we remove the argument from the exporter or do we add openqasm-pygments to the optionals list. Personally I would be in favor of just removing the argument I think.

jakelishman reviewed (commented):
My intent about pygments had been to solve the problem by entirely removing `QuantumCircuit.qasm`. `qiskit.qasm2.dumps` never had the `formatted` option, so there's nothing to deprecate.

jakelishman commented:
The other choice about Pygments in the near term is to change it to look up `qasm2` by lexer alias, and fallback to unformatted text if there's not one present.  That way, if `openqasm-pygments` is installed, it'll all just work (because it registers with the Pygments entry points), but doesn't mean that Qiskit has a dependency on any particular lexer.

mtreinish commented:
I'll just drop `QuantumCircuit.qasm` as part of this PR, it seems the simplest.

mtreinish committed:
Remove QuantumCircuit.qasm() exporter method

This commit removes the QuantumCircuit.qasm() method as part of the
legacy parser removal. The exporter method was tied to some of the
internals in the now removed `qiskit.qasm` module. The qasm() method was
already queued up for eventual removal, this commit just pushes that
forward so we don't have to add more comptaibility code into it to
keep it working as is without the qiskit.qasm module existing anymore.

mtreinish committed:
Remove jupyter qasm widget

mtreinish committed:
Update tests

mtreinish committed:
Merge branch 'main' into remove-legacy-qasm2

mtreinish committed:
Remove BlueprintCircuit .qasm() override

mtreinish committed:
Fix rst syntax in release note

jakelishman commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -1619,65 +1618,6 @@ def decompose(
         # do not copy operations, this is done in the conversion with circuit_to_dag
         return dag_to_circuit(dag, copy_operations=False)
 
-    def qasm(
-        self,
-        formatted: bool = False,
-        filename: str | None = None,
-        encoding: str | None = None,
-    ) -> str | None:
```
We need to remember to follow up this removal with a 0.46 deprecation if we're going ahead with it for 1.0 (which I think we should).

jakelishman commented on releasenotes/notes/remove-legacy-qasm2-parser-53ad3f1817fd68cc.yaml:
```diff
@@ -0,0 +1,34 @@
+    :meth:`.QuantumCircuit.from_qasm_file` or
+    :meth:`.QuantumCircuit.from_qasm_str` constructor methods and then
+    converting that :class:`.QuantumCircuit` into a :class:`.DAGCircuit` with
+    :func:`.circuit_to_dag`.
```
```suggestion
    The legacy OpenQASM 2 parser module previously present in ``qiskit.qasm`` has
    been removed. It was marked as deprecated in Qiskit 0.46.0 release. The
    OpenQASM 2 parser has been superseded by the :mod:`qiskit.qasm2` module which
    provides a faster more correct parser for QASM2.
  - |
    The ``qiskit.converters.ast_to_dag`` function has been removed. It
    previously was used to convert the abstract syntax tree generated by the
    legacy OpenQASM 2 parser (in the ``qiskit.qasm`` module which no longer exists)
    and convert that directly to a :class:`.DAGCircuit`. This function was
    marked as deprecated in the Qiskit 0.46.0 release. As the legacy
    OpenQASM 2 parser has been removed this function no longer serves a purpose.
    If you were previously using this, you can instead parse your OpenQASM 2 files
    into a :class:`.QuantumCircuit` using the
    :meth:`.QuantumCircuit.from_qasm_file` or
    :meth:`.QuantumCircuit.from_qasm_str` constructor methods and then
    converting that :class:`.QuantumCircuit` into a :class:`.DAGCircuit` with
    :func:`.circuit_to_dag`.
```

jakelishman commented on requirements.txt:
```diff
@@ -1,6 +1,5 @@
 rustworkx>=0.13.0
 numpy>=1.17,<2
-ply>=3.10
```
Very pleased to see this going, especially since it's not been maintained as an actual Python package since 2020.

mtreinish replied:
This was actually my motivation for this PR, I was looking at the requirements list and forgot this was in there.

jakelishman commented on releasenotes/notes/remove-legacy-qasm2-parser-53ad3f1817fd68cc.yaml:
```diff
@@ -0,0 +1,34 @@
+    function was marked as deprecated in the 0.46.0 release. If you were using
+    the ``QuantumCircuit.qasm()`` method to generate pygments formatted output
+    you should instead look at the standalone ``openqasm-pygments`` package
+    to provide this functionality.
```
```suggestion
  - |
    Removed the ``QuantumCircuit.qasm()`` method to generate a OpenQASM 2
    representation of the :class:`.QuantumCircuit` object. Instead the
    :func:`.qasm2.dump` or :func:`.qasm2.dumps` functions should be used. This
    function was marked as deprecated in the 0.46.0 release. If you were using
    the ``QuantumCircuit.qasm()`` method to generate pygments formatted output
    you should instead look at the standalone ``openqasm-pygments`` package
    to provide this functionality.
```

jakelishman reviewed (commented):
Thanks for this - good to get this all cleaned up.  We unified the terminology around OQ2 in #9351, so we probably should match that in the release note.

mtreinish committed:
Apply suggestions from code review

Co-authored-by: Jake Lishman <jake@binhbar.com>

1ucian0 commented:
The removal broke Randomized tests https://github.com/Qiskit/qiskit/issues/2645#issuecomment-1833062890 . Probably an easy fix.

jakelishman commented:
Yeah, the randomised tests just need to switch to using `qasm2.dumps` instead.
